consensus/src/lib.rs

```diff
@@ -78,165 +70,318 @@ pub trait PolkadotClient: Clone + 'static {
+		relay_chain.clone(),
+		announce_block,
+	);
+	let follow_finalized_head = follow_finalized_head(para_id, parachain, relay_chain);
+	select! {
+		r = follow_new_best.fuse() => r,
+		r = follow_finalized_head.fuse() => r,
+	}
 }
 
 /// Follow the relay chain new best head, to update the Parachain new best head.
-fn follow_new_best<L, P, Block, B>(
+async fn follow_new_best<P, R, Block, B>(
 	para_id: ParaId,
-	local: Arc<L>,
-	polkadot: P,
+	parachain: Arc<P>,
+	relay_chain: R,
 	announce_block: Arc<dyn Fn(Block::Hash, Vec<u8>) + Send + Sync>,
-) -> ClientResult<impl Future<Output = ()> + Send + Unpin>
+) -> ClientResult<()>
 where
 	Block: BlockT,
-	L: Finalizer<Block, B> + UsageProvider<Block> + Send + Sync + BlockBackend<Block>,
-	for<'a> &'a L: BlockImport<Block>,
-	P: PolkadotClient,
+	P: Finalizer<Block, B>
+		+ UsageProvider<Block>
+		+ Send
+		+ Sync
+		+ BlockBackend<Block>
+		+ BlockchainEvents<Block>,
+	for<'a> &'a P: BlockImport<Block>,
+	R: RelaychainClient,
 	B: Backend<Block>,
 {
-	Ok(polkadot
-		.new_best_heads(para_id)?
-		.filter_map(|head_data| {
-			let res = match <<Block as BlockT>::Header>::decode(&mut &head_data[..]) {
-				Ok(header) => Some(header),
-				Err(err) => {
-					warn!(
-						target: "cumulus-consensus",
-						"Could not decode Parachain header: {:?}", err);
-					None
-				}
-			};
-
-			future::ready(res)
-		})
-		.for_each(move |h| {
-			let hash = h.hash();
-
-			if local.usage_info().chain.best_hash == hash {
-				trace!(
-					target: "cumulus-consensus",
-					"Skipping set new best block, because block `{}` is already the best.",
-					hash,
-				)
-			} else {
-				// Make sure the block is already known or otherwise we skip setting new best.
-				match local.block_status(&BlockId::Hash(hash)) {
-					Ok(BlockStatus::InChainWithState) => {
-						// Make it the new best block
-						let mut block_import_params =
-							BlockImportParams::new(BlockOrigin::ConsensusBroadcast, h);
-						block_import_params.fork_choice = Some(ForkChoiceStrategy::Custom(true));
-						block_import_params.import_existing = true;
-
-						if let Err(err) =
-							(&*local).import_block(block_import_params, Default::default())
-						{
-							warn!(
-								target: "cumulus-consensus",
-								"Failed to set new best block `{}` with error: {:?}",
-								hash, err
-							);
-						}
-
-						(*announce_block)(hash, Vec::new());
-					}
-					Ok(BlockStatus::InChainPruned) => {
-						error!(
-							target: "cumulus-collator",
-							"Trying to set pruned block `{:?}` as new best!",
-							hash,
+	let mut new_best_heads = relay_chain.new_best_heads(para_id)?.fuse();
+	let mut imported_blocks = parachain.import_notification_stream().fuse();
+	// The unset best header of the parachain. Will be `Some(_)` when we have imported a relay chain
+	// block before the parachain block it included. In this case we need to wait for this block to
+	// be imported to set it as new best.
+	let mut unset_best_header = None;
```

Comment:
Down the line, we may even need to trigger an availability request to recover the PoV. It is possible that the initial collator is withholding the block.

Reply:
I thought that we don't wanted to prevent to ask the relay chain validators to recover the povblocks?

But yeah, in general you are right. However this is not yet implemented anywhere.